refactor(polys): simplify the DMP class

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Follows gh-25651

#### Brief description of what is fixed or changed

Simplify the DMP class. Remove the ring attribute that is only sometimes set.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
refactor(polys): simplify the DMP class

oscarbenjamin commented on sympy/polys/domains/tests/test_polynomialring.py:
```diff
@@ -25,12 +25,12 @@ def test_globalring():
-    assert X + y == X + Y == R.convert(x + y) == x + Y
-    assert X - y == X - Y == R.convert(x - y) == x - Y
+    #assert X * y == X * Y == R.convert(x * y) == x * Y
+    #assert X + y == X + Y == R.convert(x + y) == x + Y
+    #assert X - y == X - Y == R.convert(x - y) == x - Y
     assert X + 1 == R.convert(x + 1)
     raises(ExactQuotientFailed, lambda: X/Y)
-    raises(ExactQuotientFailed, lambda: x/Y)
+    #raises(ExactQuotientFailed, lambda: x/Y)
```
These tests assert mixed operations between DMP and Expr that should probably be disallowed.

oscarbenjamin commented on sympy/polys/polyclasses.py:
```diff
@@ -539,9 +539,11 @@ def exquo(f, g):
-        if f.ring and res not in f.ring:
-            from sympy.polys.polyerrors import ExactQuotientFailed
-            raise ExactQuotientFailed(f, g, f.ring)
+        if f.ring:
+            if res not in f.ring:
+                raise RuntimeError("exquo...")
+                from sympy.polys.polyerrors import ExactQuotientFailed
+                raise ExactQuotientFailed(f, g, f.ring)
```
I think that these are the only places where the `DMP.ring` attribute is used for anything nontrivial. I want to remove them all and then remove the ring attribute and then simplify `DMP.unify` and remove `per`.

oscarbenjamin commented:
@brandonwillard looks like the latest release of NumPy broke Aesara or something:
```
Run bin/test_optional_dependencies.py
ImportError while loading conftest '/home/runner/work/sympy/sympy/conftest.py'.
conftest.py:16: in <module>
    collect_ignore = _get_doctest_blacklist()
sympy/testing/runtests.py:712: in _get_doctest_blacklist
    if import_module('aesara') is None:
sympy/external/importtools.py:145: in import_module
    mod = __import__(module, **import_kwargs)
/opt/hostedtoolcache/Python/3.[11](https://github.com/sympy/sympy/actions/runs/6210142565/job/16858063332?pr=25691#step:14:12).5/x64/lib/python3.11/site-packages/aesara/__init__.py:[12](https://github.com/sympy/sympy/actions/runs/6210142565/job/16858063332?pr=25691#step:14:13)0: in <module>
    from aesara import scalar, tensor
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/tensor/__init__.py:106: in <module>
    from aesara.tensor import (  # noqa
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/tensor/blas.py:162: in <module>
    from aesara.tensor.blas_headers import blas_header_text, blas_header_version
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/tensor/blas_headers.py:1015: in <module>
    if not config.blas__ldflags:
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/configparser.py:354: in __get__
    val_str = self.default()
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/link/c/cmodule.py:2725: in default_blas_ldflags
    blas_info = np.__config__.get_info("blas_opt")
E   AttributeError: module 'numpy.__config__' has no attribute 'get_info'
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/configparser.py", line 234, in fetch_val_for_key
    return self._aesara_cfg.get(section, option)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/configparser.py", line 797, in get
    d = self._unify_values(section, vars)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/configparser.py", line 1168, in _unify_values
    raise NoSectionError(section) from None
configparser.NoSectionError: No section: 'blas'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/configparser.py", line 350, in __get__
    val_str = cls.fetch_val_for_key(self.name, delete_key=delete_key)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/configparser.py", line 238, in fetch_val_for_key
    raise KeyError(key)
KeyError: 'blas__ldflags'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/runner/work/sympy/sympy/sympy/testing/runtests.py", line 797, in _doctest
    blacklist.extend(_get_doctest_blacklist())
                     ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/sympy/sympy/sympy/testing/runtests.py", line 712, in _get_doctest_blacklist
    if import_module('aesara') is None:
       ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/sympy/sympy/sympy/external/importtools.py", line [14](https://github.com/sympy/sympy/actions/runs/6210142565/job/16858063332?pr=25691#step:14:15)5, in import_module
    mod = __import__(module, **import_kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/__init__.py", line 120, in <module>
    from aesara import scalar, tensor
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/tensor/__init__.py", line 106, in <module>
    from aesara.tensor import (  # noqa
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/tensor/blas.py", line 162, in <module>
    from aesara.tensor.blas_headers import blas_header_text, blas_header_version
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/tensor/blas_headers.py", line 10[15](https://github.com/sympy/sympy/actions/runs/6210142565/job/16858063332?pr=25691#step:14:16), in <module>
    if not config.blas__ldflags:
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/configparser.py", line 354, in __get__
    val_str = self.default()
              ^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/aesara/link/c/cmodule.py", line 27[25](https://github.com/sympy/sympy/actions/runs/6210142565/job/16858063332?pr=25691#step:14:26), in default_blas_ldflags
    blas_info = np.__config__.get_info("blas_opt")
                ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'numpy.__config__' has no attribute 'get_info'
```

brandonwillard commented:
Thanks; I'll check it out!

oscarbenjamin committed:
fix(holonomic): Convert Expr to domain explicitly

oscarbenjamin committed:
fix(polys): reduce dependence of DMP on .ring

oscarbenjamin committed:
fix(polys): reduce dependence of DMF on .ring

oscarbenjamin committed:
fix(polys): reduce dependence of DMF on .ring

oscarbenjamin committed:
refactor(polys): set DMP.ring to None always

oscarbenjamin commented:
@jksuom (or anyone else) do you have any idea what `GeneralizedPolynomialRing` is supposed to be:
https://github.com/sympy/sympy/blob/28b834a2f73e352f8dee82932b5ad504f5d5fbe7/sympy/polys/domains/old_polynomialring.py#L317-L318
If the objects are `DMF` then that implies that they are rational functions but it somehow does not behave like rational functions.

Creating an old polynomial ring with `order=ilex` creates these:
https://github.com/sympy/sympy/blob/28b834a2f73e352f8dee82932b5ad504f5d5fbe7/sympy/polys/domains/old_polynomialring.py#L466-L469

There are tests like this that don't make any sense:
https://github.com/sympy/sympy/blob/28b834a2f73e352f8dee82932b5ad504f5d5fbe7/sympy/polys/domains/tests/test_polynomialring.py#L57
If `X` and `Y` are `DMF` then this division should succeed.

Currently this requires checks like this:
https://github.com/sympy/sympy/blob/28b834a2f73e352f8dee82932b5ad504f5d5fbe7/sympy/polys/polyclasses.py#L1395-L1398

I want to remove `f.ring` so this would not work. I don't understand though what `GeneralizedPolynomialRing` is supposed to be and why it is not the same as `GlobalPolynomialRing`.

Here `f.ring` is:
```python
In [9]: R = QQ.old_poly_ring(x, y, order='ilex')

In [10]: R.convert(x)
Out[10]: DMF(([[mpq(1,1)], []], [[mpq(1,1)]]), QQ, QQ[x,y order=ilex])

In [11]: R.convert(x).ring
Out[11]: QQ[x,y order=ilex]

In [12]: type(R.convert(x).ring)
Out[12]: sympy.polys.domains.old_polynomialring.GeneralizedPolynomialRing
```

Why does it use `DMF` if the elements are supposed to be "polynomials"?

jksuom commented:
A "generalized polynomial ring" is a [localization](https://en.wikipedia.org/wiki/Localization_(commutative_algebra)

oscarbenjamin committed:
fix(printing): do not use DMP.ring in printing

oscarbenjamin commented:
> It is a subring of the fraction field so its elements are DMF. In SymPy, the denominators are nonzero at the origin so the fractions are well defined there.

Thanks @jksuom.

So far in this PR I have just disabled the membership check in `f / g` and apart from specific tests for `ExactQuotientFailed` all tests involving this are passing.

So this is a ring and not a field in which case `X / Y` can be disallowed.

The membership testing could be moved from DMF to `GeneralisedPolynomialRing.exquo` which can handle checking if the division is contained in the ring.  Then `DMF.__truediv__` can just be defined as it should for rational functions.

Any user of this ring should then use e.g. `R.exquo(f, g)` rather than `f / g` as is the case for `ZZ` and other non-field rings. Then `R.exquo` would just compute `f / g` using `DMF` and then test for membership afterwards.

Does that seem correct to you?

oscarbenjamin committed:
fix(holonomic): pass domain explicitly

oscarbenjamin committed:
fix(holonomic): fix printing of holonomic

jksuom commented:
> and then test for membership afterwards. Does that seem correct to you?

Yes, it seems that it would suffice to check the leading monomial of the denominator.

oscarbenjamin committed:
fix(polys): fix printing related to DMP/DMF

oscarbenjamin committed:
fix(polys): add exquo to GeneralizedPolynomialring

oscarbenjamin commented:
> > and then test for membership afterwards. Does that seem correct to you?
> 
> Yes, it seems that it would suffice to check the leading monomial of the denominator.

I've restored the check in `exquo` by calling `.new`:
https://github.com/sympy/sympy/blob/9a20bc0bfc10424ae48b419253974a1e875b237f/sympy/polys/domains/old_polynomialring.py#L360-L371
The check in .new is:
https://github.com/sympy/sympy/blob/9a20bc0bfc10424ae48b419253974a1e875b237f/sympy/polys/domains/old_polynomialring.py#L324-L331
Does that seem correct?

oscarbenjamin commented on doc/src/modules/polys/domainsintro.rst:
```diff
@@ -925,7 +925,7 @@ two versions of :ref:`K[x]` using :py:meth:`~.Domain.poly_ring` and
   x**2 + 1
   >>> p2
-  x**2 + 1
+  DMP([1, 0, 1], ZZ, None)
```
Since `DMP` no longer has a `.ring` attribute it does not have any generators and cannot print as e.g. `x**2 + 1`.

Any code that just prints a raw `DMF` will not show the raw representation. Previously DMF would sometimes have a ring and sometimes not making it a  bit unreliable for printing and many other things.

oscarbenjamin commented on sympy/holonomic/holonomic.py:
```diff
@@ -2647,17 +2649,16 @@ def _extend_y0(Holonomic, n):
         return frac.diff()
 
-    K = frac.ring
     p = K.numer(frac)
```
We cannot get the ring from the `DMF` directly. Instead the ring must be passed explicitly if it is needed.

oscarbenjamin commented on sympy/polys/domains/quotientring.py:
```diff
@@ -28,7 +28,8 @@ def __init__(self, ring, data):
 
     def __str__(self):
         from sympy.printing.str import sstr
-        return sstr(self.data) + " + " + str(self.ring.base_ideal)
+        data = self.ring.ring.to_sympy(self.data)
+        return sstr(data) + " + " + str(self.ring.base_ideal)
```
In some cases the old printing behaviour can be recovered at a higher-level by using the ring to convert the `DMP` into an expression with symbols.

oscarbenjamin commented on sympy/polys/domains/tests/test_polynomialring.py:
```diff
@@ -23,20 +23,16 @@ def test_globalring():
-    raises(ExactQuotientFailed, lambda: x/Y)
-    raises(ExactQuotientFailed, lambda: X/y)
+    raises(TypeError, lambda: x/Y)
+    raises(TypeError, lambda: X/y)
```
Since `DMP` no longer has `.ring` it cannot interact directly with Expr like this (it does not know what its generators are).

oscarbenjamin committed:
refactor(polys): remove DMP.ring and DMF.ring

jksuom commented:
The check in `new` looks OK assuming that the monomial order (e.g. ilex) works as expected.

oscarbenjamin committed:
fix: remove unneeded imports

oscarbenjamin committed:
refactor(polys): remove kill param

oscarbenjamin committed:
refactor(polys): remove optional per args

oscarbenjamin committed:
polys: remove implicit conversions from per

oscarbenjamin committed:
fix(polys): remove implicit conversions in DMP

oscarbenjamin committed:
fix(polys): remove duplicate DMP.from_dict method

oscarbenjamin commented:
I think that this is a good milestone so this is ready for review.

What this does:

Remove the ring attribute from both DMP and DMF. These types would sometimes have the ring attribute but sometimes not and so there were a lot of conditionals in the code but it was not handled correctly everywere. The changes here mean that DMP and DMF never have a ring attribute. The implication of this is that a DMP is more like PurePoly in that its generators are unspecified.

Removing the ring attribute means that a DMP must always print like `DMP([1, 2], ZZ)` whereas previously if it had the ring attribute it could print more like `x + 2`. In many cases I could patch up the printing code at a higher level to include the generators but in some I couldn't and some doctests needed to be changed to show the `DMF(...)` type output.

The constructors for `DMP` are simplified. The optional parameters for `per` were removed and the implicit conversions performed in the `DMP` constructor were removed and replaced with explicit conversions like `DMP.from_dict` or `DMP.ground_new`.

I think I have preserved the behaviour of GeneralizedPolynomialring but I have removed the checks from `a / b` and moved them to `R.exquo(a, b)` instead. This is consistent without how it works for other rings like `ZZ`.

These changes follow on from gh-25651. The intention is that the `DMP` class should be simplified enough that we can create an alternate implementation of it based on python-flint's `fmpz_poly` etc types that would make things like polynomial factorisation much more efficient for univariate polynomials.

Longer term when it is possible to use Flint's sparse polynomials via python-flint I would like to add them here as well and also it probably makes sense to have an implementation based on SymPy's existing sparse polynomials. This implies that there would be 4 implementations of the DMP interface, something like:

- DMP (SymPy's dense multivariate polynomials)
- SMP? (SymPy's sparse multivariate polynomials - wraps PolyElement)
- FUP? (Flint univariate dense polynomials)
- FMP? (Flint multivariate sparse polynomials)

Needing to implement the interface this many times means that we need the interface to be kept as simple and well defined as possible. I have removed some niceties here like mixing Expr with DMP in arithmetic and some other conversions because they should not really be needed in the lower-level API that `DMP` is intended for.

Further simplifications are possible but I would defer those for not to get this in as the main change and then consider other changes at the time of trying to implement a wrapper for Flint's univariate matrices.

oscarbenjamin commented:
The optional dependencies job will fail in CI because of the Aesara problem mentioned above.

oscarbenjamin commented:
> The optional dependencies job will fail in CI because of the Aesara problem mentioned above.

Any PR can still be merged if that job fails though.

oscarbenjamin committed:
perf(polys): disable runtime type checks in DMP

oscarbenjamin commented:
Thanks!
